### PR TITLE
Copy generated CRDs by kustomize to Helm

### PR DESCRIPTION
**Pull request by Jooho**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #https://github.com/kserve/kserve/issues/3388

**(NOTE)** 
This needs to solve [this ticket](https://github.com/kserve/kserve/issues/3344) first. because the CRD from kustomize from KServe has kserve own information like conversion webook. 

**Type of changes**
Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [X] Test A

~~~
git clone https://github.com/Jooho/kserve.git
cd kserve
./hack/generate-install.sh  v0.12.0-rc1
~~~

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
Could you do a diff check to see the manifest now matches?

**Review by terrytangyuan — APPROVED** [review]
/lgtm

File: hack/generate-install.sh [hunk 1]
```diff
@@ -74,10 +74,30 @@ kubectl kustomize config/clusterresources | sed s/:latest/:$TAG/ > $RUNTIMES_INS
 # Update ingressGateway in inferenceservice configmap as 'kubeflow/kubeflow-gateway'
 yq -i 'select(.metadata.name == "inferenceservice-config").data.ingress |= (fromjson | .ingressGateway = "kubeflow/kubeflow-gateway" | tojson)' $KUBEFLOW_INSTALL_PATH
 
+# Create a temp directory for final CRD
+mkdir config/crd_helm
+delimeter_lines=$(cat ${INSTALL_PATH} -n |grep '\-\-\-'|cut -f1)
```
Comment (ckadner) [inline hack/generate-install.sh]: This works on Linux but does not work on Mac OS. Can you swap the `-n` and the `${INSTALL_PATH}`?


```sh
$ INSTALL_PATH=./install/v0.12.0-rc1/kserve.yaml

$ cat ${INSTALL_PATH} -n |grep '\-\-\-'|cut -f1

cat: -n: No such file or directory
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---
---


[kserve] ((v0.12.0-rc1) *)$ cat -n ${INSTALL_PATH} | grep '\-\-\-' | cut -f1

     9
  1904
  2568
  3099
 18815
 20710
 20821
 20831
 20873
 21178
 21196
 21210
 21223
 21236
 21585
 21591
 21612
 21629
 21641
 21731
 21745
 21753
 21812
 21842
 21872
 21902
 21932
```

File: hack/generate-install.sh [hunk 2]
```diff
@@ -74,10 +74,30 @@ kubectl kustomize config/clusterresources | sed s/:latest/:$TAG/ > $RUNTIMES_INS
 # Update ingressGateway in inferenceservice configmap as 'kubeflow/kubeflow-gateway'
 yq -i 'select(.metadata.name == "inferenceservice-config").data.ingress |= (fromjson | .ingressGateway = "kubeflow/kubeflow-gateway" | tojson)' $KUBEFLOW_INSTALL_PATH
 
+# Create a temp directory for final CRD
+mkdir config/crd_helm
```
Comment (ckadner) [inline hack/generate-install.sh]: this one failed on my second attempt where the `config/crd_helm` directory was created (but not deleted) on the previous run.  Might be safer to include a `-p`

```
mkdir -p config/crd_helm
```

File: hack/generate-install.sh [hunk 3]
```diff
@@ -74,10 +74,30 @@ kubectl kustomize config/clusterresources | sed s/:latest/:$TAG/ > $RUNTIMES_INS
...
+done
+
 # Copy CRD files to charts crds directory
-cp config/crd/serving.kserve.io_clusterservingruntimes.yaml charts/kserve-crd/templates/serving.kserve.io_clusterservingruntimes.yaml
-cp config/crd/serving.kserve.io_inferenceservices.yaml charts/kserve-crd/templates/serving.kserve.io_inferenceservices.yaml
-cp config/crd/serving.kserve.io_trainedmodels.yaml charts/kserve-crd/templates/serving.kserve.io_trainedmodels.yaml
-cp config/crd/serving.kserve.io_inferencegraphs.yaml charts/kserve-crd/templates/serving.kserve.io_inferencegraphs.yaml
-cp config/crd/serving.kserve.io_servingruntimes.yaml charts/kserve-crd/templates/serving.kserve.io_servingruntimes.yaml
-cp config/crd/serving.kserve.io_clusterstoragecontainers.yaml charts/kserve-crd/templates/serving.kserve.io_clusterstoragecontainers.yaml
+cp config/crd_helm/serving.kserve.io_clusterservingruntimes.yaml charts/kserve-crd/templates/serving.kserve.io_clusterservingruntimes.yaml
```
Comment (ckadner) [inline hack/generate-install.sh]: did you not want to include the updated CRDs in this PR?

Comment (Jooho) [reply hack/generate-install.sh]: No, I don't want to add them.

File: hack/generate-install.sh (hunk 1 above)
Comment (Jooho) [reply hack/generate-install.sh]: updated

File: hack/generate-install.sh (hunk 2 above)
Comment (Jooho) [reply hack/generate-install.sh]: updated

**Comment by Jooho** [comment]
@ckadner could you please trigger to retest the failed test?
/retest

File: hack/generate-install.sh (hunk 3 above)
Comment (ckadner) [reply hack/generate-install.sh]: without seeing the diff here, how would those changes be tested/verified? At the time the next KServe release (candidate)?

**Comment by Jooho** [comment]
@terrytangyuan @ckadner please review this pr :) looks it is ready to be merged

**Comment by ckadner** [comment]
> @terrytangyuan @ckadner please review this pr :) looks it is ready to be merged

Your script appears to work as intended on MacOS now, although a bit hard to read. But since the results of running the script (the changed helm charts) are not part of this PR, how would those changes be tested/verified? At the time the next KServe release (candidate)? Maybe @israel-hdez could help verify as part of the next MM release?

image

**Comment by Jooho** [comment]
The object of the change is simple.
- Generated all manifests by Kustomize.
- Check the big 1 file and check each manifest kind/name based on `---` 
- If a manifest is CRD that we are looking for, then it will be saved into a separate file.
- Then the separated files will be copied to Helm folder.

It is best to test this script while a new build is in progress.

**Comment by yuzisun** [comment]
We have done code generation verification in CI for generated go client
https://github.com/kserve/kserve/blob/master/hack/verify-codegen.sh

**Comment by sivanantha321** [comment]
> We have done code generation verification in CI for generated go client https://github.com/kserve/kserve/blob/master/hack/verify-codegen.sh

Actually We are not using this script anymore. We just re run the generate command and use git diff to check for any difference. https://github.com/kserve/kserve/blob/cefdc2ecff9764b318d975f1bfa31e71c6874362/.github/workflows/verify-codegen.yml#L38

**Comment by Jooho** [comment]
@yuzisun @sivanantha321 so how do we need to verify this?

File: hack/generate-install.sh [hunk 4]
```diff
@@ -77,10 +77,31 @@ kubectl kustomize config/clusterresources | sed s/:latest/:$TAG/ > $RUNTIMES_INS
 # Update ingressGateway in inferenceservice configmap as 'kubeflow/kubeflow-gateway'
 yq -i 'select(.metadata.name == "inferenceservice-config").data.ingress |= (fromjson | .ingressGateway = "kubeflow/kubeflow-gateway" | tojson)' $KUBEFLOW_INSTALL_PATH
 
+# Create a temp directory for final CRD
+temp_dir=/tmp/charts
```
Comment (spolti) [inline hack/generate-install.sh]: mktemp is safer

**Comment by spolti** [comment]
I am little bit confused.

How
```bash
cp config/crd/full/serving.kserve.io_clusterservingruntimes.yaml charts/kserve-crd/templates/serving.kserve.io_clusterservingruntimes.yaml
```

is different than:
```bash
cp ${temp_dir}/serving.kserve.io_clusterservingruntimes.yaml charts/kserve-crd/templates/serving.kserve.io_clusterservingruntimes.yaml
```
?

File: hack/generate-install.sh (hunk 4 above)
Comment (Jooho) [reply hack/generate-install.sh]: updated

**Comment by Jooho** [comment]
@spolti previously it copied the original CRD to the chart but there are some patches in the config folder such as conversion webhook.  So it makes some differences between HELM and Kustomize. 
If the user uses `kustomize`, they use patched CRD but if the user uses `helm` it does not have those patches. So we should use the generated CRD by kustomize for helm chart and modelmesh as well to use the same CRDs

**Comment by spolti** [comment]
> @spolti previously it copied the original CRD to the chart but there are some patches in the config folder such as conversion webhook. So it makes some differences between HELM and Kustomize. If the user uses `kustomize`, they use patched CRD but if the user uses `helm` it does not have those patches. So we should use the generated CRD by kustomize for helm chart and modelmesh as well to use the same CRDs

I see, thanks for the explanation.

**Pull request merged**